treegp/two_pcf.py

```diff
@@ -0,0 +1,410 @@
+            g1 = np.linspace(-0.3, 0.3, N_restart)
+            size = np.linspace(p0[0] - p0[0]/10., 2*p0[0], N_restart)
+            g1, g2, size = np.meshgrid(g1, g1, size)
+            N_restart = int(N_restart**3)
+            g1 = g1.reshape(N_restart)
+            g2 = g2.reshape(N_restart)
+            size = size.reshape(N_restart)
+            for i in range(N_restart):
+                print("restart fit because failure")
+                new_p0 = [size[i], g1[i], g2[i]]
+                print(new_p0)
+                self._minimize_minuit(p0=new_p0)
+                if self._fit_ok:
+                    break
+        pcf = self._model_skl(self.result[0], self.result[1], 
+                              self.result[2], self.result[3])
+
+class two_pcf(object):
+    """
+    Fit statistical uncertaintie on two-point correlation function using bootstraping.
+
+    :param X:           Coordinates of the field.  (n_samples, 1 or 2)
+    :param y:           Values of the field. (n_samples)
+    :param y_err:       Error of y. (n_samples)
+    :param min_sep:     Minimum bin for treecorr. (float)
+    :param max_sep:     Maximum bin for treecorr. (float)
+    :param nbins:       Number of bins (if 1D correlation function) of the square root of the number
+                        of bins (if 2D correlation function) used in TreeCorr to compute the
+                        2-point correlation function. [default: 20]
+    :param anisotropic: 2D 2-point correlation function.
+                        Used 2D correlation function for the
+                        fiting part of the GP. (Boolean)
+    :param robust_fit:  Used minuit to fit hyperparameter. Works only
+                        anisotropic is True. (Boolean)
+    """
+    def __init__(self, X, y, y_err,
+                 min_sep, max_sep, nbins=20, 
+                 anisotropic=False, robust_fit=False, 
+                 p0=[3000., 0., 0.]):
+        self.ndim = np.shape(X)[1]
+        if self.ndim not in [1, 2]:
+            raise ValueError('two-pcf support only 1d and 2d modeling for the moment. curent ndim: %i'%(self.ndim))
+        if self.ndim == 2:
+            self.X = X
+        if self.ndim == 1:
+            self.X = np.array([X.T, np.zeros_like(X.T)]).T[:,0]
+        self.y = y
+        self.y_err = y_err
+        self.min_sep = min_sep
+        self.max_sep = max_sep
+        self.nbins = nbins
+        self.anisotropic = anisotropic
+        self.robust_fit = robust_fit
+        self.p0_robust_fit = p0
+
+    def resample_bootstrap(self):
+        """
+        Make a single bootstrap resampling on data.
+        """
+        npsfs = len(self.y)
+
+        ind_object = np.random.randint(0,npsfs-1, size=npsfs)
+        u_ressample = self.X[:,0][ind_object]
+        v_ressample = self.X[:,1][ind_object]
+        y_ressample = self.y[ind_object]
+        y_err_ressample = self.y_err[ind_object]
+
+        return u_ressample, v_ressample, y_ressample, y_err_ressample
```

Comment:
It may be more straightforward to use the [SciPy bootstrap function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html).

Reply:
I guess the full covariance matrix computation should be rewrite. When I wrote that, `treecorr` was not doing it and now this is the case: [treecorr covariance matrix ](https://rmjarvis.github.io/TreeCorr/_build/html/cov.html#covariance-matrix)